src/java/org/apache/cassandra/service/accord/txn/TxnUpdate.java

```diff
@@ -0,0 +1,157 @@
+import org.apache.cassandra.io.IVersionedSerializer;
+import org.apache.cassandra.io.util.DataInputPlus;
+import org.apache.cassandra.io.util.DataOutputPlus;
+import org.apache.cassandra.service.accord.SerializationUtils;
+import org.apache.cassandra.transport.ProtocolVersion;
+import org.apache.cassandra.utils.ByteBufferUtil;
+import org.apache.cassandra.utils.ObjectSizes;
+
+import static org.apache.cassandra.service.accord.SerializationUtils.deserialize;
+import static org.apache.cassandra.service.accord.SerializationUtils.serialize;
+import static org.apache.cassandra.service.accord.SerializationUtils.serializeArray;
+import static org.apache.cassandra.utils.ByteBufferUtil.readWithVIntLength;
+import static org.apache.cassandra.utils.ByteBufferUtil.serializedSizeWithVIntLength;
+import static org.apache.cassandra.utils.ByteBufferUtil.writeWithVIntLength;
+
+public class TxnUpdate implements Update
+{
+    private static final long EMPTY_SIZE = ObjectSizes.measure(new TxnUpdate(new ByteBuffer[0], null));
+
+    private final ByteBuffer[] serializedUpdates;
+    private final ByteBuffer serializedCondition;
+
+    public TxnUpdate(List<TxnWrite.Fragment> updates, TxnCondition condition)
+    {
+        this.serializedUpdates = serialize(updates, TxnWrite.Fragment.serializer);
+        this.serializedCondition = serialize(condition, TxnCondition.serializer);
+    }
+
+    public TxnUpdate(ByteBuffer[] serializedUpdates, ByteBuffer serializedCondition)
+    {
+        this.serializedUpdates = serializedUpdates;
+        this.serializedCondition = serializedCondition;
+    }
+
+    public ByteBuffer serializedCondition()
+    {
+        return serializedCondition.duplicate();
+    }
+
+    public long estimatedSizeOnHeap()
+    {
+        long size = EMPTY_SIZE + ByteBufferUtil.estimatedSizeOnHeap(serializedCondition);
+        for (ByteBuffer update : serializedUpdates)
+            size += ByteBufferUtil.estimatedSizeOnHeap(update);
+        return size;
+    }
+
+    @Override
+    public String toString()
+    {
+        return "TxnUpdate{" +
+               "updates=" + deserialize(serializedUpdates, TxnWrite.Fragment.serializer) +
+               ", condition=" + deserialize(serializedCondition, TxnCondition.serializer) +
+               '}';
+    }
+
+    @Override
+    public boolean equals(Object o)
+    {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        TxnUpdate txnUpdate = (TxnUpdate) o;
+        return Arrays.equals(serializedUpdates, txnUpdate.serializedUpdates) && Objects.equals(serializedCondition, txnUpdate.serializedCondition);
+    }
+
+    @Override
+    public int hashCode()
+    {
+        int result = Objects.hash(serializedCondition);
+        result = 31 * result + Arrays.hashCode(serializedUpdates);
+        return result;
+    }
+
+    @Override
+    public Keys keys()
+    {
+        // TODO: Does this actually work?
```

Comment:
It doesn't I think. I would expect no writes to be applied. This eventually filters to [here](https://github.com/apache/cassandra-accord/blob/197e4c740e4e755fe477ecc0b6f2284df253d37a/accord-core/src/main/java/accord/primitives/Writes.java#L79) and left folds over no keys not applying the writes.